--- wget-1.19.4/testenv/conf/expected_files.py.org	2018-01-30 09:36:46.359579482 +0100
+++ wget-1.19.4/testenv/conf/expected_files.py	2018-01-30 09:37:01.350209794 +0100
@@ -24,9 +24,9 @@ class ExpectedFiles:
         snapshot = {}
         for parent, dirs, files in os.walk('.'):
             for name in files:
-                # pubring.kbx, dirmngr.conf, gpg.conf will be created by libgpgme if $HOME doesn't contain the .gnupg directory.
+                # pubring.gpg, pubring.kbx, dirmngr.conf, gpg.conf will be created by libgpgme if $HOME doesn't contain the .gnupg directory.
                 # setting $HOME to CWD (in base_test.py) breaks two Metalink tests, so we skip this file here.
-                if name in [ 'pubring.kbx', 'dirmngr.conf', 'gpg.conf' ]:
+                if name in [ 'pubring.gpg', 'pubring.kbx', 'dirmngr.conf', 'gpg.conf' ]:
                     continue
 
                 f = {'content': ''}
