diff -Nurb wget-1.9.1.bef/src/http.c wget-1.9.1.new/src/http.c
--- wget-1.9.1.bef/src/http.c	2005-05-29 09:12:59.000000000 +0200
+++ wget-1.9.1.new/src/http.c	2005-05-29 12:43:11.000000000 +0200
@@ -818,8 +818,8 @@
     }
   else
     range = NULL;
-  if (opt.useragent)
-    STRDUP_ALLOCA (useragent, opt.useragent);
+  if (opt.user_agent)
+    STRDUP_ALLOCA (useragent, opt.user_agent);
   else
     {
       useragent = (char *)alloca (10 + strlen (version_string));
diff -Nurb wget-1.9.1.bef/src/init.c wget-1.9.1.new/src/init.c
--- wget-1.9.1.bef/src/init.c	2005-05-29 09:12:59.000000000 +0200
+++ wget-1.9.1.new/src/init.c	2005-05-29 12:44:38.000000000 +0200
@@ -218,7 +218,7 @@
   { "timestamping",	&opt.timestamping,	cmd_boolean },
   { "tries",		&opt.ntry,		cmd_number_inf },
   { "useproxy",		&opt.use_proxy,		cmd_boolean },
-  { "useragent",	NULL,			cmd_spec_useragent },
+  { "useragent",	&opt.user_agent,	cmd_spec_useragent },
   { "verbose",		&opt.verbose,		cmd_boolean },
   { "wait",		&opt.wait,		cmd_time },
   { "waitretry",	&opt.waitretry,		cmd_time }
@@ -1203,7 +1203,7 @@
 	       exec_name, com, val);
       return 0;
     }
-  opt.useragent = xstrdup (val);
+  opt.user_agent = xstrdup (val);
   return 1;
 }
 
@@ -1345,7 +1345,7 @@
   FREE_MAYBE (opt.https_proxy);
   FREE_MAYBE (opt.http_proxy);
   free_vec (opt.no_proxy);
-  FREE_MAYBE (opt.useragent);
+  FREE_MAYBE (opt.user_agent);
   FREE_MAYBE (opt.referer);
   FREE_MAYBE (opt.http_user);
   FREE_MAYBE (opt.http_passwd);
diff -Nurb wget-1.9.1.bef/src/options.h wget-1.9.1.new/src/options.h
--- wget-1.9.1.bef/src/options.h	2005-05-29 09:12:59.000000000 +0200
+++ wget-1.9.1.new/src/options.h	2005-05-29 12:39:53.000000000 +0200
@@ -137,7 +137,7 @@
   int backup_converted;		/* Do we save pre-converted files as *.orig? */
   int backups;			/* Are numeric backups made? */
 
-  char *useragent;		/* Naughty User-Agent, which can be
+  char *user_agent;		/* Naughty User-Agent, which can be
 				   set to something other than
 				   Wget. */
   char *referer;		/* Naughty Referer, which can be
