diff -Nur wget-1.8.2.orig/acinclude.m4 wget-1.8.2/acinclude.m4
--- wget-1.8.2.orig/acinclude.m4	Thu Jan  1 01:00:00 1970
+++ wget-1.8.2/acinclude.m4	Wed Jun  5 20:45:08 2002
@@ -0,0 +1,11 @@
+AC_DEFUN(WGET_STRUCT_UTIMBUF,
+[AC_MSG_CHECKING(for struct utimbuf)
+if test x"$ac_cv_header_utime_h" = xyes; then
+  AC_EGREP_CPP([struct[ 	]+utimbuf],
+    [#include <utime.h>],
+    [AC_DEFINE(HAVE_STRUCT_UTIMBUF,,[Have struct utimbuf])
+      AC_MSG_RESULT(yes)],
+    AC_MSG_RESULT(no))
+else
+  AC_MSG_RESULT(no)
+fi])
diff -Nur wget-1.8.2.orig/configure.in wget-1.8.2/configure.in
--- wget-1.8.2.orig/configure.in	Sat May 18 05:05:11 2002
+++ wget-1.8.2/configure.in	Wed Jun  5 20:44:55 2002
@@ -46,14 +46,14 @@
 dnl Get cannonical host
 dnl
 AC_CANONICAL_HOST
-AC_DEFINE_UNQUOTED(OS_TYPE, "$host_os")
+AC_DEFINE_UNQUOTED(OS_TYPE, "$host_os", [OS type])
 
 dnl
 dnl Process features.
 dnl
 AC_ARG_WITH(socks,
 [  --with-socks            use the socks library],
-[AC_DEFINE(HAVE_SOCKS)])
+[AC_DEFINE(HAVE_SOCKS,,[Use the socks library])])
 
 AC_ARG_WITH(ssl,
 [  --with-ssl[=SSL-ROOT]   link with SSL support [default=auto]
@@ -62,17 +62,17 @@
 AC_ARG_ENABLE(opie,
 [  --disable-opie          disable support for opie or s/key FTP login],
 USE_OPIE=$enableval, USE_OPIE=yes)
-test x"${USE_OPIE}" = xyes && AC_DEFINE(USE_OPIE)
+test x"${USE_OPIE}" = xyes && AC_DEFINE(USE_OPIE,,[FTP opie or s/key login support])
 
 AC_ARG_ENABLE(digest,
 [  --disable-digest        disable support for HTTP digest authorization],
 USE_DIGEST=$enableval, USE_DIGEST=yes)
-test x"${USE_DIGEST}" = xyes && AC_DEFINE(USE_DIGEST)
+test x"${USE_DIGEST}" = xyes && AC_DEFINE(USE_DIGEST,,[HTTP digest authorization support])
 
 AC_ARG_ENABLE(debug,
 [  --disable-debug         disable support for debugging output],
 DEBUG=$enableval, DEBUG=yes)
-test x"${DEBUG}" = xyes && AC_DEFINE(DEBUG)
+test x"${DEBUG}" = xyes && AC_DEFINE(DEBUG,,[Support for debugging output])
 
 wget_need_md5=no
 
@@ -374,7 +374,7 @@
     dnl AC_MSG_RESULT doesn't look right here, but I'm not sure what
     dnl to use instead.
     AC_MSG_RESULT("Compiling in support for SSL in $ssl_root")
-    AC_DEFINE(HAVE_SSL)
+    AC_DEFINE(HAVE_SSL,,[SSL support])
     AC_SUBST(SSL_INCLUDES)
     SSL_OBJ='gen_sslfunc$o'
     AC_SUBST(SSL_OBJ)
@@ -416,7 +416,7 @@
   dnl specific to the Solaris MD5 library.
   if test x"$found_md5" = xno; then
     AC_CHECK_LIB(md5, md5_calc, [
-      AC_DEFINE(HAVE_SOLARIS_MD5)
+      AC_DEFINE(HAVE_SOLARIS_MD5,,[Use Solaris MD5])
       LIBS="-lmd5 $LIBS"
       found_md5=yes
     ])
@@ -426,19 +426,19 @@
   dnl implementation.
   if test x"$found_md5" = xno; then
     if test x"$ssl_success" = xyes; then
-      AC_DEFINE(HAVE_OPENSSL_MD5)
+      AC_DEFINE(HAVE_OPENSSL_MD5,,[Use OpenSSL MD5])
       found_md5=yes
     fi
   fi
 
   dnl If none of the above worked, use the one we ship with Wget.
   if test x"$found_md5" = xno; then
-    AC_DEFINE(HAVE_BUILTIN_MD5)
+    AC_DEFINE(HAVE_BUILTIN_MD5,,[Use builtin MD5])
     found_md5=yes
     MD5_OBJ="$MD5_OBJ gnu-md5\$o"
   fi
 fi
-AC_DEFINE(HAVE_MD5)
+AC_DEFINE(HAVE_MD5,,[Have MD5])
 AC_SUBST(MD5_OBJ)
 
 dnl
@@ -452,7 +452,7 @@
 dnl new language was added.
 
 dnl internationalization macros
-WGET_WITH_NLS
+AM_GNU_GETTEXT
 
 dnl
 dnl Find makeinfo.  If makeinfo is not found, look for Emacs.  If
diff -Nur wget-1.8.2.orig/src/main.c wget-1.8.2/src/main.c
--- wget-1.8.2.orig/src/main.c	Wed Jun  5 20:39:23 2002
+++ wget-1.8.2/src/main.c	Wed Jun  5 20:51:34 2002
@@ -44,11 +44,11 @@
 #ifdef HAVE_SIGNAL_H
 # include <signal.h>
 #endif
-#ifdef HAVE_NLS
+#ifdef ENABLE_NLS
 #ifdef HAVE_LOCALE_H
 # include <locale.h>
 #endif /* HAVE_LOCALE_H */
-#endif /* HAVE_NLS */
+#endif /* ENABLE_NLS */
 #include <errno.h>
 
 #define OPTIONS_DEFINED_HERE	/* for options.h */
@@ -97,9 +97,9 @@
 static void
 i18n_initialize (void)
 {
-  /* If HAVE_NLS is defined, assume the existence of the three
+  /* If ENABLE_NLS is defined, assume the existence of the three
      functions invoked here.  */
-#ifdef HAVE_NLS
+#ifdef ENABLE_NLS
   /* Set the current locale.  */
   /* Here we use LC_MESSAGES instead of LC_ALL, for two reasons.
      First, message catalogs are all of I18N Wget uses anyway.
@@ -116,7 +116,7 @@
   /* Set the text message domain.  */
   bindtextdomain ("wget", LOCALEDIR);
   textdomain ("wget");
-#endif /* HAVE_NLS */
+#endif /* ENABLE_NLS */
 }
 
 /* Print the usage message.  */
diff -Nur wget-1.8.2.orig/src/wget.h wget-1.8.2/src/wget.h
--- wget-1.8.2.orig/src/wget.h	Sat May 18 05:14:49 2002
+++ wget-1.8.2/src/wget.h	Wed Jun  5 20:50:56 2002
@@ -53,14 +53,12 @@
 
 /* `gettext (FOO)' is long to write, so we use `_(FOO)'.  If NLS is
    unavailable, _(STRING) simply returns STRING.  */
-#ifdef HAVE_NLS
+#ifdef ENABLE_NLS
 # define _(string) gettext (string)
-# ifdef HAVE_LIBINTL_H
 #  include <libintl.h>
-# endif /* HAVE_LIBINTL_H */
-#else  /* not HAVE_NLS */
+#else  /* not ENABLE_NLS */
 # define _(string) string
-#endif /* not HAVE_NLS */
+#endif /* not ENABLE_NLS */
 
 /* I18N NOTE: You will notice that none of the DEBUG messages are
    marked as translatable.  This is intentional, for a few reasons:
