--- wget-1.7/configure.in~	Tue Aug 21 23:18:24 2001
+++ wget-1.7/configure.in	Tue Aug 21 23:17:12 2001
@@ -88,8 +88,7 @@
 	for i in inria kame linux toshiba v6d zeta; do
 		case $i in
 		inria)
-			dnl http://www.kame.net/
-			AC_EGREP_CPP(yes, [dnl
+			AC_EGREP_CPP(yes, [
 #include <netinet/in.h>
 #ifdef IPV6_INRIA_VERSION
 yes
@@ -98,8 +97,7 @@
 				CFLAGS="-DINET6 $CFLAGS"])
 			;;
 		kame)
-			dnl http://www.kame.net/
-			AC_EGREP_CPP(yes, [dnl
+			AC_EGREP_CPP(yes, [
 #include <netinet/in.h>
 #ifdef __KAME__
 yes
@@ -108,7 +106,6 @@
 				CFLAGS="-DINET6 $CFLAGS"])
 			;;
 		linux)
-			dnl http://www.v6.linux.or.jp/
 			if test -d /usr/inet6; then
 				ipv6type=$i
 				ipv6lib=inet6
@@ -117,7 +114,7 @@
 			fi
 			;;
 		toshiba)
-			AC_EGREP_CPP(yes, [dnl
+			AC_EGREP_CPP(yes, [
 #include <sys/param.h>
 #ifdef _TOSHIBA_INET6
 yes
@@ -128,7 +125,7 @@
 				CFLAGS="-DINET6 $CFLAGS"])
 			;;
 		v6d)
-			AC_EGREP_CPP(yes, [dnl
+			AC_EGREP_CPP(yes, [
 #include </usr/local/v6/include/sys/v6config.h>
 #ifdef __V6D__
 yes
@@ -139,7 +136,7 @@
 				CFLAGS="-I/usr/local/v6/include $CFLAGS"])
 			;;
 		zeta)
-			AC_EGREP_CPP(yes, [dnl
+			AC_EGREP_CPP(yes, [
 #include <sys/param.h>
 #ifdef _ZETA_MINAMI_INET6
 yes
