diff -uNr wget-1.8.2.orig/NEWS wget-1.8.2/NEWS
--- wget-1.8.2.orig/NEWS	Thu Aug 28 21:45:31 2003
+++ wget-1.8.2/NEWS	Thu Aug 28 21:45:31 2003
@@ -232,7 +232,7 @@
 
 ** Compiles on pre-ANSI compilers.
 
-** Global wgetrc now goes to /usr/local/etc (i.e. $sysconfdir).
+** Global wgetrc now goes to /etc (i.e. $sysconfdir).
 
 ** Lots of bugfixes.
 
@@ -295,9 +295,8 @@
 ** Fixed a long-standing bug, so that Wget now works over SLIP
 connections.
 
-** You can have a system-wide wgetrc (/usr/local/lib/wgetrc by
-default). Settings in $HOME/.wgetrc override the global ones, of
-course :-)
+** You can have a system-wide wgetrc (/etc/wgetrc by default). Settings
+in $HOME/.wgetrc override the global ones, of course :-)
 
 ** You can set up quota in .wgetrc to prevent sucking too much
 data. Try `quota = 5M' in .wgetrc (or quota = 100K if you want your
diff -uNr wget-1.8.2.orig/doc/wget.info-1 wget-1.8.2/doc/wget.info-1
--- wget-1.8.2.orig/doc/wget.info-1	Thu Aug 28 21:48:44 2003
+++ wget-1.8.2/doc/wget.info-1	Thu Aug 28 21:48:44 2003
@@ -115,8 +115,7 @@
    * Most of the features are fully configurable, either through
      command line options, or via the initialization file `.wgetrc'
      (*note Startup File::).  Wget allows you to define "global"
-     startup files (`/usr/local/etc/wgetrc' by default) for site
-     settings.
+     startup files (`/etc/wgetrc' by default) for site settings.
 
 
    * Finally, GNU Wget is free software.  This means that everyone may
diff -uNr wget-1.8.2.orig/doc/wget.info-2 wget-1.8.2/doc/wget.info-2
--- wget-1.8.2.orig/doc/wget.info-2	Thu Aug 28 21:49:57 2003
+++ wget-1.8.2/doc/wget.info-2	Thu Aug 28 21:49:57 2003
@@ -521,9 +521,7 @@
 ===============
 
    When initializing, Wget will look for a "global" startup file,
-`/usr/local/etc/wgetrc' by default (or some prefix other than
-`/usr/local', if Wget was not installed there) and read commands from
-there, if it exists.
+`/etc/wgetrc' by default and read commands from there, if it exists.
 
    Then it will look for the user's file.  If the environmental variable
 `WGETRC' is set, Wget will try to load that file.  Failing that, no
@@ -533,7 +531,7 @@
 
    The fact that user's settings are loaded after the system-wide ones
 means that in case of collision user's wgetrc _overrides_ the
-system-wide wgetrc (in `/usr/local/etc/wgetrc' by default).  Fascist
+system-wide wgetrc (in `/etc/wgetrc' by default).  Fascist
 admins, away!
 
 
@@ -873,7 +871,7 @@
      ## not contain a comprehensive list of commands -- look at the manual
      ## to find out what you can put into this file.
      ##
-     ## Wget initialization file can reside in /usr/local/etc/wgetrc
+     ## Wget initialization file can reside in /etc/wgetrc
      ## (global, for all users) or $HOME/.wgetrc (for a single user).
      ##
      ## To use the settings in this file, you will have to uncomment them,
@@ -882,7 +880,7 @@
      
      
      ##
-     ## Global settings (useful for setting up in /usr/local/etc/wgetrc).
+     ## Global settings (useful for setting up in /etc/wgetrc).
      ## Think well before you change them, since they may reduce wget's
      ## functionality, and make it behave contrary to the documentation:
      ##
diff -uNr wget-1.8.2.orig/doc/wget.texi wget-1.8.2/doc/wget.texi
--- wget-1.8.2.orig/doc/wget.texi	Thu Aug 28 21:52:07 2003
+++ wget-1.8.2/doc/wget.texi	Thu Aug 28 21:52:07 2003
@@ -195,12 +195,12 @@
 Most of the features are fully configurable, either through command line
 options, or via the initialization file @file{.wgetrc} (@pxref{Startup
 File}).  Wget allows you to define @dfn{global} startup files
-(@file{/usr/local/etc/wgetrc} by default) for site settings.
+(@file{/etc/wgetrc} by default) for site settings.
 
 @ignore
 @c man begin FILES
 @table @samp
-@item /usr/local/etc/wgetrc
+@item /etc/wgetrc
 Default location of the @dfn{global} startup file.
 
 @item .wgetrc
@@ -1920,9 +1920,7 @@
 @cindex location of wgetrc
 
 When initializing, Wget will look for a @dfn{global} startup file,
-@file{/usr/local/etc/wgetrc} by default (or some prefix other than
-@file{/usr/local}, if Wget was not installed there) and read commands
-from there, if it exists.
+@file{/etc/wgetrc} by default and read commands from there, if it exists.
 
 Then it will look for the user's file.  If the environmental variable
 @code{WGETRC} is set, Wget will try to load that file.  Failing that, no
@@ -1932,7 +1930,7 @@
 
 The fact that user's settings are loaded after the system-wide ones
 means that in case of collision user's wgetrc @emph{overrides} the
-system-wide wgetrc (in @file{/usr/local/etc/wgetrc} by default).
+system-wide wgetrc (in @file{/etc/wgetrc} by default).
 Fascist admins, away!
 
 @node Wgetrc Syntax, Wgetrc Commands, Wgetrc Location, Startup File
