--- wget-1.19.2/testenv/conf/expected_files.py.org	2017-10-30 12:36:46.911716601 +0100
+++ wget-1.19.2/testenv/conf/expected_files.py	2017-10-30 12:41:03.358656484 +0100
@@ -24,9 +24,9 @@ class ExpectedFiles:
         snapshot = {}
         for parent, dirs, files in os.walk('.'):
             for name in files:
-                # pubring.kbx will be created by libgpgme if $HOME doesn't contain the .gnupg directory.
+                # pubring.kbx, dirmngr.conf, gpg.conf can be created by libgpgme if $HOME doesn't contain the .gnupg directory.
 					 # setting $HOME to CWD (in base_test.py) breaks two Metalink tests, so we skip this file here.
-                if name == 'pubring.kbx':
+                if name == 'pubring.kbx' or name == 'dirmngr.conf' or name == 'gpg.conf':
                     continue
 
                 f = {'content': ''}
