--- wget-1.10.1/configure.in.orig	2005-08-11 23:35:27.000000000 +0200
+++ wget-1.10.1/configure.in	2005-08-28 13:42:34.625493608 +0200
@@ -437,6 +437,7 @@
 ALL_LINGUAS=`(cd ${srcdir}/po && ls *.po | sed -e 's/\.po$//' | tr '\012' ' ')`
 
 dnl internationalization macros
+AM_GNU_GETTEXT
 WGET_WITH_NLS
 
 dnl
@@ -473,5 +474,4 @@
 #include "config-post.h"
 ])
 AC_CONFIG_FILES([stamp-h], [echo timestamp > stamp-h])
-AC_CONFIG_COMMANDS([default], [WGET_PROCESS_PO])
 AC_OUTPUT
diff -urN wget-1.9-b5.org/Makefile.in wget-1.9-b5/Makefile.in
--- wget-1.9-b5.org/Makefile.in	2003-10-12 23:55:54.000000000 +0200
+++ wget-1.9-b5/Makefile.in	2003-10-13 00:29:14.000000000 +0200
@@ -82,7 +82,7 @@
 manext='$(manext)'
 
 # subdirectories in the distribution
-SUBDIRS = src doc po tests util windows
+SUBDIRS = src doc intl po tests util windows
 
 # default target
 all: libtool src/config.h Makefile $(SUBDIRS)
@@ -106,7 +106,7 @@
 
 # Install `.mo' files
 install.mo:
-	cd po && $(MAKE) $(MAKEDEFS) $@
+	cd po && $(MAKE) $(MAKEDEFS) install
 
 # create tag files for Emacs
 TAGS:
--- wget-1.10/src/main.c.orig	2005-06-22 03:25:18.000000000 +0200
+++ wget-1.10/src/main.c	2005-06-22 03:25:48.000000000 +0200
@@ -43,7 +43,7 @@
 #endif /* HAVE_UNISTD_H */
 #include <string.h>
 #include <signal.h>
-#ifdef HAVE_NLS
+#ifdef ENABLE_NLS
 # include <locale.h>
 #endif
 #include <assert.h>
@@ -90,13 +90,13 @@
 static void
 i18n_initialize (void)
 {
-  /* HAVE_NLS implies existence of functions invoked here.  */
-#ifdef HAVE_NLS
+  /* ENABLE_NLS implies existence of functions invoked here.  */
+#ifdef ENABLE_NLS
   /* Set the current locale.  */
   setlocale (LC_ALL, "");
   /* Set the text message domain.  */
   bindtextdomain ("wget", LOCALEDIR);
   textdomain ("wget");
-#endif /* HAVE_NLS */
+#endif /* ENABLE_NLS */
 }
 
 /* Definition of command-line options. */
--- wget-1.10/src/wget.h.orig	2005-06-22 03:26:51.000000000 +0200
+++ wget-1.10/src/wget.h	2005-06-22 03:27:04.000000000 +0200
@@ -50,20 +50,20 @@
 
 /* `gettext (FOO)' is long to write, so we use `_(FOO)'.  If NLS is
    unavailable, _(STRING) simply returns STRING.  */
-#ifdef HAVE_NLS
+#ifdef ENABLE_NLS
 # define _(string) gettext (string)
 # ifdef HAVE_LIBINTL_H
 #  include <libintl.h>
 # else  /* not HAVE_LIBINTL_H */
    const char *gettext ();
 # endif /* not HAVE_LIBINTL_H */
-#else  /* not HAVE_NLS */
+#else  /* not ENABLE_NLS */
 # define _(string) (string)
 # define ngettext(sing, plur, num)  ((num) == 1 ? (sing) : (plur))
 # undef HAVE_WCHAR_H
 # undef HAVE_WCWIDTH
 # undef HAVE_MBTOWC
-#endif /* not HAVE_NLS */
+#endif /* not ENABLE_NLS */

 #if HAVE_WCWIDTH && HAVE_MBTOWC
 # define USE_NLS_PROGRESS_BAR 1
