--- wget-1.9.1.orig/src/main.c	2004-12-28 01:42:49.000000000 +0200
+++ wget-1.9.1/src/main.c	2004-12-28 01:42:01.000000000 +0200
@@ -380,6 +380,17 @@
 
   append_to_log = 0;
 
+#ifdef ENABLE_IPV6
+  {
+      int s = socket(AF_INET6, SOCK_STREAM, 0);
+      if (s < 0 && (errno == EAFNOSUPPORT || errno == EINVAL)) {
+	  ip_default_family = AF_INET;
+      } else {
+	  close(s);
+      }
+  }
+#endif
+
   /* Construct the name of the executable, without the directory part.  */
   exec_name = strrchr (argv[0], PATH_SEPARATOR);
   if (!exec_name)
diff -urN wget-1.9.1.org/src/ftp-basic.c wget-1.9.1/src/ftp-basic.c
--- wget-1.9.1.org/src/ftp-basic.c	2003-11-08 20:17:55.000000000 +0100
+++ wget-1.9.1/src/ftp-basic.c	2005-04-25 00:08:41.000000000 +0200
@@ -276,6 +276,9 @@
     /* Huh?  This is not BINDERR! */
     return BINDERR;
   inet_ntop (AF_INET6, &in_addr, ipv6, sizeof (ipv6));
+  
+  if (strncmp(ipv6, "::ffff:", 7) == 0)
+      memmove(ipv6, ipv6 + 7, sizeof(ipv6) - 7);
 
   /* Construct the argument of EPRT (of the form |2|IPv6.ascii|PORT.ascii|). */
   bytes = alloca (3 + strlen (ipv6) + 1 + numdigit (port) + 1 + 1);
