--- wget-1.14/doc/texi2pod.pl.orig	2013-12-06 20:29:19.785649931 +0100
+++ wget-1.14/doc/texi2pod.pl	2013-12-06 20:32:09.288982037 +0100
@@ -291,7 +291,11 @@
 	if (defined $1) {
             my $thing = $1;
             if ($ic =~ /\@asis/) {
-                $_ = "\n=item $thing\n";
+		if ($thing =~ /^[0-9]+$/) {
+		    $_ = "\n=item C\&LT;$thing\&GT;\n";
+		} else {
+		    $_ = "\n=item $thing\n";
+		}
             } else {
                 # Entity escapes prevent munging by the <> processing below.
                 $_ = "\n=item $ic\&LT;$thing\&GT;\n";
