--- wget-1.9.1/configure.in.orig	2003-11-08 20:15:35.000000000 +0100
+++ wget-1.9.1/configure.in	2003-11-16 11:46:45.947950144 +0100
@@ -541,7 +541,8 @@
 dnl new language was added.
 
 dnl internationalization macros
-WGET_WITH_NLS
+AM_GNU_GETTEXT
+WGET_WITH_NLS
 
 dnl
 dnl Find makeinfo.  We used to provide support for Emacs processing
diff -urN wget-1.9-b5.org/Makefile.in wget-1.9-b5/Makefile.in
--- wget-1.9-b5.org/Makefile.in	2003-10-12 23:55:54.000000000 +0200
+++ wget-1.9-b5/Makefile.in	2003-10-13 00:29:14.000000000 +0200
@@ -82,7 +82,7 @@
 infodir='$(infodir)' mandir='$(mandir)' manext='$(manext)'
 
 # subdirectories in the distribution
-SUBDIRS = src doc po util windows
+SUBDIRS = src doc intl po util windows
 
 # default target
 all: libtool src/config.h Makefile $(SUBDIRS)
@@ -106,7 +106,7 @@
 
 # Install `.mo' files
 install.mo:
-	cd po && $(MAKE) $(MAKEDEFS) $@
+	cd po && $(MAKE) $(MAKEDEFS) install
 
 # Regenerate libtool if ltconfig and/or ltmain are updated
 libtool: $(LIBTOOL_DEPS)
--- wget-1.10/src/main.c.orig	2005-06-22 03:25:18.000000000 +0200
+++ wget-1.10/src/main.c	2005-06-22 03:25:48.000000000 +0200
@@ -43,11 +43,11 @@
 #ifdef HAVE_SIGNAL_H
 # include <signal.h>
 #endif
-#ifdef HAVE_NLS
+#ifdef ENABLE_NLS
 #ifdef HAVE_LOCALE_H
 # include <locale.h>
 #endif /* HAVE_LOCALE_H */
-#endif /* HAVE_NLS */
+#endif /* ENABLE_NLS */
 #include <assert.h>
 
 #include <errno.h>
@@ -90,8 +90,8 @@
 static void
 i18n_initialize (void)
 {
-  /* HAVE_NLS implies existence of functions invoked here.  */
-#ifdef HAVE_NLS
+  /* ENABLE_NLS implies existence of functions invoked here.  */
+#ifdef ENABLE_NLS
   /* Set the current locale.  */
   /* Where possible, sets only LC_MESSAGES and LC_CTYPE.  Other
      categories, such as numeric, time, or collation, break code that
@@ -110,7 +110,7 @@
   /* Set the text message domain.  */
   bindtextdomain ("wget", LOCALEDIR);
   textdomain ("wget");
-#endif /* HAVE_NLS */
+#endif /* ENABLE_NLS */
 }
 
 /* Definition of command-line options. */
--- wget-1.10/src/wget.h.orig	2005-06-22 03:26:51.000000000 +0200
+++ wget-1.10/src/wget.h	2005-06-22 03:27:04.000000000 +0200
@@ -50,16 +50,16 @@
 
 /* `gettext (FOO)' is long to write, so we use `_(FOO)'.  If NLS is
    unavailable, _(STRING) simply returns STRING.  */
-#ifdef HAVE_NLS
+#ifdef ENABLE_NLS
 # define _(string) gettext (string)
 # ifdef HAVE_LIBINTL_H
 #  include <libintl.h>
 # else  /* not HAVE_LIBINTL_H */
    const char *gettext ();
 # endif /* not HAVE_LIBINTL_H */
-#else  /* not HAVE_NLS */
+#else  /* not ENABLE_NLS */
 # define _(string) (string)
-#endif /* not HAVE_NLS */
+#endif /* not ENABLE_NLS */
 
 /* A pseudo function call that serves as a marker for the automated
    extraction of messages, but does not call gettext().  The run-time
